io_scene_godot/converters/material_script_lib.py

```diff
@@ -0,0 +1,643 @@
+"""Prewritten shader scripts used to export cycles material"""
+import re
+
+FUNCTION_HEAD_PATTERN = re.compile(
+    (r'void\s+([a-zA-Z]\w*)\s*\(((\s*(out\s+)?'
+     r'(vec2|vec3|vec4|float|mat4|sampler2D)\s+[a-zA-Z]\w*\s*,?)*)\)'),
+)
+
+
+class ScriptsFunction:
+    """Structure caches information about a parsed shader function"""
+    def __init__(self, code, input_sockets=(), output_sockets=()):
+        # at most one group
+        self.code = code
+        self.in_param_types = list()
+        self.out_param_types = list()
+
+        # linked socket ids of material node
+        self.in_sockets = input_sockets
+        self.out_sockets = output_sockets
+
+        matched_group = FUNCTION_HEAD_PATTERN.findall(code)[0]
+        self.name = matched_group[0]
+        parameters_str = matched_group[1]
+
+        for param_str in parameters_str.strip().split(','):
+            tokens = tuple([x.strip() for x in param_str.split()])
+            if tokens[0] == 'out':
+                self.out_param_types.append(tokens[1])
+            else:
+                self.in_param_types.append(tokens[0])
+
+
+NODE_SCRIPTS = dict()
+
+# supported material node in shader node category
+SHADER_NODES = {
+    'ShaderNodeBsdfPrincipled',
+    'ShaderNodeEmission',
+    'ShaderNodeBsdfDiffuse',
+    'ShaderNodeBsdfGlossy',
+    'ShaderNodeBsdfGlass',
+}
+
+NODE_SCRIPTS['ShaderNodeBsdfPrincipled'] = ScriptsFunction(
+    code="""
+void pincipled_bsdf_node(vec4 color, float subsurface, vec4 subsurface_color,
+        float specular, float metallic, float roughness, float clearcoat,
+        float clearcoat_roughness, float anisotropy, float transmission,
+        float IOR, vec3 view_tangent, vec3 view_normal, out vec4 color_out,
+        out float sss_strength_out, out float metallic_out,
+        out float specular_out, out float roughness_out, out float
+        clearcoat_out, out float clearcoat_gloss_out, out float anisotropy_out,
+        out float transmission_out, out float eta, out vec3 tangent_out,
+        out vec3 normal_out) {
+    metallic = clamp(metallic, 0.0, 1.0);
+    transmission = clamp(transmission, 0.0, 1.0);
+
+    subsurface = subsurface * (1.0 - metallic);
+    transmission = transmission * (1.0 - metallic);
+
+    color_out = mix(color, subsurface_color, subsurface);
+    sss_strength_out = subsurface;
+    metallic_out = metallic;
+    specular_out = pow((IOR - 1.0)/(IOR + 1.0), 2)/0.08;
+    roughness_out = clamp(roughness, 1e-8, 0.9999);
+    clearcoat_out = clearcoat * (1.0 - transmission);
+    clearcoat_gloss_out = 1.0 - clearcoat_roughness;
+    anisotropy_out = clamp(anisotropy, 0.0, 1.0);
+    transmission_out = transmission;
+    eta = 1.0 / IOR;
+    tangent_out = view_tangent;
+    normal_out = view_normal;
+}
+""",
+    input_sockets=[
```

Comment:
Possibly use tuples instead of lists? Ideally these shouldn't change, and tuples would prevent accidental changes

Reply:
hah, actually, I initial use tuples and then switch to use list.  The problem of tuples is that for one item tuple, you need a additional comma (e.g. ( Value, ) ), it is easy to forget